ext/opcache/jit/zend_jit_x86.dasc

```diff
@@ -14233,7 +14233,11 @@ static int zend_jit_hash_jmp(dasm_State **Dst, const zend_op *opline, const zend
 	}
 	|	LOAD_ADDR FCARG1a, jumptable
 	|	sub r0, aword [FCARG1a + offsetof(HashTable, arData)]
-	|	mov FCARG1a, (sizeof(Bucket) / sizeof(void*))
+	if (HT_IS_PACKED(jumptable)) {
+		|	mov FCARG1a, (sizeof(zval) / sizeof(void*))
```

Comment:
Not related to this PR, noticed while reading: On x64, sizeof(Bucket)=32 sizeof(zval)=16, and `void*` is a power of two on both. Elsewhere, I see the optimization shl/shr used instead of dividing or multiplying by a power of 2, e.g. right above this.

On x86, sizeof(zval) is also 16.

```
					|.if X64
						|	mov r0, FCARG2a
						|	shl r0, 4
					|.else
						|	imul r0, FCARG2a, sizeof(zval)
					|.endif
```

So I think everything except 32-bit Buckets can be micro-optimized for a `switch` by using `shr` instead of `idiv FCARG1a` below to save a mov and a few cycles?

Reply:
Would it be easier to read and write code (or for others to contribute, or to emit an assertion error if struct sizes changed) if there were macros for these magic numbers?

It may also allow avoiding the need to repeat `ZEND_ASSERT(sizeof(Bucket) == 32);` and similar assertions for other structures if the macro itself was what contained the assertion.

For example,

`MULTIPLY_REGISTER_BY_POWER_OF_TWO FCARG2a, divisor, dividend`
`MULTIPLY_REGISTER_BY_POWER_OF_TWO FCARG2a, sizeof(zval), 1`
`DIVIDE_REGISTER_BY_POWER_OF_TWO sizeof(zval), sizeof(void*)`

(maybe omit the smaller_value)

The function `zend_hash_check_size` links to `BitScanReverse` and `__builtin_clz` macros for efficiently doing that division - I'm not sure which operations common C compilers can optimize out at runtime, though `__builtin_constant_p` may help omit the assertions if anything actually can't be

e.g. I mean something along the lines of `#define fast_get_log2(size) __builtin_constant_p(size) ? ((size) == 1 ? 0 : (size == 2 ? 1) ? (size == 4 ? 2) : ordinary_get_log2(size)) : ordinary_get_log2(size)`

Reply:
I tried replacing multiplication by shift, but unfortunately this didn't make any speed improvement even on bench.php

Reply:
> I tried replacing multiplication by shift, but unfortunately this didn't make any speed improvement even on bench.php

Should bench.php be updated to include at least one benchmarks of match or switch statements, or maybe another benchmark that's used for less common operations?

It looks like this code is in the function `static int zend_jit_hash_jmp`, where the only caller is `static int zend_jit_switch()`, which is used for `match` expressions and `switch` statements

```
					case ZEND_SWITCH_LONG:
					case ZEND_SWITCH_STRING:
					case ZEND_MATCH:
						if (!zend_jit_switch(&dasm_state, opline, op_array, ssa, NULL, NULL)) {
							goto jit_failure;
						}
						goto done;
```

`» grep switch Zend/bench.php` returned zero results. Same for `match`.

Reply:
It's better not to change bench.php.
Micro benchmarks for switch/match may go into micro_bench.php